classlib/src/main/java/org/teavm/classlib/java/util/TFormatter.java

```diff
@@ -240,11 +242,70 @@ private void formatValue(char specifier) throws IOException {
                     formatRadixInt(specifier, 4, true);
                     break;
 
+                case 'f':
+                    formatFloat(specifier, false);
+                    break;
+
                 default:
                     throw new UnknownFormatConversionException(String.valueOf(specifier));
             }
         }
 
+        private void formatFloat(char specifier, boolean upperCase) throws IOException {
+            verifyFlags(specifier, MASK_FOR_INT_DECIMAL_FORMAT);
+            verifyFloatFlags();
+
+            TDecimalFormat format = new TDecimalFormat();
+            format.setMaximumIntegerDigits(width);
+            format.setMaximumFractionDigits(precision);
+            format.setDecimalFormatSymbols(new DecimalFormatSymbols(locale));
+            format.setGroupingUsed((flags & TFormattableFlags.GROUPING_SEPARATOR) != 0);
+            if ((flags & TFormattableFlags.PARENTHESIZED_NEGATIVE) != 0) {
+                format.setNegativePrefix("(");
+                format.setNegativeSuffix(")");
+            }
+            if ((flags & TFormattableFlags.SIGNED) != 0) {
+                format.setPositivePrefix("+"); // dfs has no plus sign
+            } else if ((flags & TFormattableFlags.LEADING_SPACE) != 0) {
+                format.setPositivePrefix(" ");
+            }
+
+            StringBuilder sb;
+            Object arg = args[argumentIndex];
+            if (arg instanceof Double) {
+                double value = (Double) arg;
+                sb = new StringBuilder(format.format(value));
+            } else if (arg instanceof Float) {
+                float value = (Float) arg;
+                sb = new StringBuilder(format.format(value));
+            } else if (arg instanceof BigDecimal) {
+                BigDecimal value = (BigDecimal) arg;
+                sb = new StringBuilder(format.format(value));
+            } else {
+                throw new IllegalFormatConversionException(specifier, arg != null ? arg.getClass() : null);
+            }
+
+            if ((flags & TFormattableFlags.ZERO_PADDED) != 0) {
+                for (int i = sb.length(); i < width; ++i) {
+                    sb.append(Character.forDigit(0, 10));
```

Comment:
This should be zero digit from DecimalFormatSymbols. But anyway, I don't like this approach. I'd rather looked for a way to make TDecimalFormat do this padding

Reply:
This has been addressed with setMinimumIntegerDigits